clients/drcachesim/scheduler/scheduler.cpp

```diff
@@ -669,16 +669,145 @@ scheduler_tmpl_t<RecordType, ReaderType>::read_recorded_schedule()
+    // See comment in read_recorded_schedule() on our assumption that we can
+    // easily fit the whole context switch sequence in memory.
+    // We do not have a subfile listing feature in archive_istream_t, but we can
+    // read sequentially as each record has a cpu field.
+    // This schedule_entry_t format doesn't have the stop instruction ordinal (as it was
+    // designed for skip targets only), so we take two passes to get that information.
+    // We also need to translate the thread and cpu id values into 0-based ordinals.
+    std::unordered_map<memref_tid_t, input_ordinal_t> tid2input;
+    for (int i = 0; i < static_cast<input_ordinal_t>(inputs_.size()); ++i) {
+        tid2input[inputs_[i].tid] = i;
+    }
+    std::unordered_map<uint64_t, output_ordinal_t> cpu2output;
+    std::vector<std::map<uint64_t, uint64_t>> start2stop(inputs_.size());
+    // We number the outputs according to their order in the file.
+    // XXX: Should we support some direction from the user on this?
+    // What about dividing by socket: we may want to do that using the cpuid.
+    output_ordinal_t cur_output = 0;
+    uint64_t cur_cpu = std::numeric_limits<uint64_t>::max();
+    // We also want to collapse same-cpu consecutive records so we start with
+    // a temporary local vector.
+    std::vector<std::vector<schedule_record_t>> all_sched(outputs_.size());
+    while (options_.replay_as_traced_istream->read(reinterpret_cast<char *>(&entry),
+                                                   sizeof(entry))) {
+        if (entry.cpu != cur_cpu) {
+            if (cur_cpu != std::numeric_limits<uint64_t>::max()) {
+                ++cur_output;
+                if (cur_output >= static_cast<int>(outputs_.size())) {
+                    error_string_ = "replay_as_traced_istream cpu count != output count";
+                    return STATUS_ERROR_INVALID_PARAMETER;
+                }
+            }
+            cur_cpu = entry.cpu;
+            cpu2output[entry.cpu] = cur_output;
+        }
+        input_ordinal_t input = tid2input[entry.thread];
+        // We'll fill in the stop ordinal in our second pass below.
+        uint64_t start = entry.instr_count;
+        uint64_t timestamp = entry.timestamp;
+        // Some entries have no instructions.
+        if (all_sched[cur_output].size() > 0 &&
+            input == all_sched[cur_output].back().key.input &&
+            start == all_sched[cur_output].back().start_instruction) {
+            VPRINT(this, 3,
+                   "Output #%d: as-read segment #%zu has no instructions: skipping\n",
+                   cur_output, all_sched[cur_output].size() - 1);
+            continue;
+        }
+        all_sched[cur_output].emplace_back(schedule_record_t::DEFAULT, input, start, 0,
+                                           timestamp);
+        start2stop[input][entry.instr_count] = entry.instr_count;
+    }
+    for (int i = 0; i < static_cast<output_ordinal_t>(outputs_.size()); ++i) {
```

Comment:
Would it help memory usage efficiency if we collapsed each output as soon as it is created, instead of reading all outputs first? Or does that make code more complex?

Reply:
It is actually impossible because each entry only stores its start and not its stop so we don't know if there was a migration in between two consecutive starts on one cpu.  The original use case for this cpu_schedule file was to make per-core skipping faster.  Maybe we could make a v2 that adds the stop; for now I would say let's live with these 2 passes and having to pull the whole every-timestamp entry into memory and only if we hit issues should we spend more time on it.  I augmented the comment toward the top of the file regarding this.

Reply:
I don't mean that we should remove the two passes. Currently, we first read all entries for *all* outputs. I was suggesting reading for one output and then collapsing it, then reading the entries for the next output.

Reply:
I guess it won't make that big of a difference since the no-instruction entries are rare.

Reply:
The no-instruction entries are already collapsed as they're read so they do not occupy memory.  The same-cpuid-consecutive entries are what are impossible to collapse until all other outputs are read.